File: app/SieveFilter.scala
import java.io.File

case class SieveFilterConfig(
  keepOnly2xx3xx: Boolean,
  ipPrefixBlockFilePath: Option[String]
)

class SieveFilter(config: SieveFilterConfig) {

  private val blockedPrefixIPs = config.ipPrefixBlockFilePath.map(path => {
    try {
      scala.io.Source.fromFile(new File(path)).getLines.toList.
        filter(_.trim.nonEmpty).
        filterNot(_.startsWith("#")).
        toSet
    } catch {
      case e: Exception => {
        println(s"error: failed to load IP prefix block file $path")
        Set.empty[String]
      }
    }
  }).getOrElse(Set.empty[String])

  def isBlocked(request: LoggedRequest): Boolean = {
    val typ = request.sc_status / 100
    (config.keepOnly2xx3xx && !Set(2, 3).contains(typ)) ||
      (request.c_ip.exists(ip => blockedPrefixIPs.exists(ip.startsWith)))
  }
}

File: app/SieveFilter.ts
import { readFileSync } from 'fs';
import { LoggedRequest } from './Model';

export interface SieveFilterConfig {
  keepOnly2xx3xx: boolean;
  ipPrefixBlockFilePath?: string;
}

export class SieveFilter {
  private blockedPrefixIPs: Set<string>;

  constructor(private config: SieveFilterConfig) {
    this.blockedPrefixIPs = new Set();

    if (config.ipPrefixBlockFilePath) {
      try {
        const content = readFileSync(config.ipPrefixBlockFilePath, 'utf-8');
        const lines = content.split('\n')
          .map(line => line.trim())
          .filter(line => line.length > 0)
          .filter(line => !line.startsWith('#'));
        
        this.blockedPrefixIPs = new Set(lines);
      } catch (e) {
        console.log(`error: failed to load IP prefix block file ${config.ipPrefixBlockFilePath}`);
        this.blockedPrefixIPs = new Set();
      }
    }
  }

  isBlocked(request: LoggedRequest): boolean {
    const statusType = Math.floor(request.sc_status / 100);
    const isWrongStatusType = this.config.keepOnly2xx3xx && ![2, 3].includes(statusType);
    
    const hasBlockedIP = request.c_ip !== undefined && 
      Array.from(this.blockedPrefixIPs).some(prefix => request.c_ip!.startsWith(prefix));

    return isWrongStatusType || hasBlockedIP;
  }
}
